main/autogen/autoloader.cc

```diff
@@ -323,23 +341,52 @@ void DefTreeBuilder::addParsedFileDefinitions(const core::GlobalState &gs, const
     }
 }
 
-void DefTree::markPackageNamespace(core::NameRef mangledName, const vector<core::NameRef> &nameParts) {
+DefTree *DefTree::findNode(const vector<core::NameRef> &nameParts) {
     DefTree *node = this;
     for (auto nr : nameParts) {
         auto it = node->children.find(nr);
         if (it == node->children.end()) {
-            return;
+            return nullptr;
         }
         node = it->second.get();
     }
-    ENFORCE(!pkgName.exists(), "Package name should not be already set");
+
+    return node;
+}
+
+void DefTree::markPackageNamespace(core::NameRef mangledName, const vector<core::NameRef> &nameParts) {
+    DefTree *node = this->findNode(nameParts);
+    if (node == nullptr) {
+        return;
+    }
+
+    ENFORCE(!(node->pkgName.exists()), "Package name should not be already set");
     node->pkgName = mangledName;
 }
 
-void DefTreeBuilder::markPackages(const core::GlobalState &gs, DefTree &root) {
+void DefTreeBuilder::markPackages(const core::GlobalState &gs, DefTree &root, const AutoloaderConfig &alCfg) {
+    auto testRoot = root.findNode({core::Names::Constants::Test()});
+
     for (auto nr : gs.packageDB().packages()) {
         auto &pkg = gs.packageDB().getPackageInfo(nr);
-        root.markPackageNamespace(pkg.mangledName(), pkg.fullName());
+        if (pkg.strictAutoloaderCompatibility()) {
+            // Only mark strictly path-based autoload compatible packages for now to reduce
+            // computation / code generation, given this is the only current use-case for registering
+            // packages in this context in the Stripe codebase.
+
+            // Additionally this package must be registed for path-based autoloading.
+            // TODO: (aadi-stripe, 10/24/2022) Remove this functionality once we no longer require
+            // special registration.
+            auto &pkgFullName = pkg.fullName();
+            if (!alCfg.registeredForPBAL(pkgFullName)) {
+                continue;
+            }
```

Comment:
Only mark `strict` compatible packages that are additionally registered for PBAL via:
```
--autogen-autoloader-pbal-namespaces Opus::Admin
```